include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,709 @@
+  template <typename Number>
+  inline
+  typename ReadWriteVector<Number>::iterator
+  ReadWriteVector<Number>::begin ()
+  {
+    return &val[0];
+  }
+
+
+
+  template <typename Number>
+  inline
+  typename ReadWriteVector<Number>::const_iterator
+  ReadWriteVector<Number>::begin () const
+  {
+    return &val[0];
+  }
+
+
+
+  template <typename Number>
+  inline
+  typename ReadWriteVector<Number>::iterator
+  ReadWriteVector<Number>::end ()
+  {
+    return (&val[this->size()-1])+1;
+  }
+
+
+
+  template <typename Number>
+  inline
+  typename ReadWriteVector<Number>::const_iterator
+  ReadWriteVector<Number>::end () const
+  {
+    return (&val[this->size()-1])+1;
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number
+  ReadWriteVector<Number>::operator() (const size_type global_index) const
+  {
+    return val[global_to_local(global_index)];
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number &
+  ReadWriteVector<Number>::operator() (const size_type global_index)
+  {
+    return val[global_to_local (global_index)];
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number
+  ReadWriteVector<Number>::operator[] (const size_type global_index) const
+  {
+    return operator()(global_index);
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number &
+  ReadWriteVector<Number>::operator[] (const size_type global_index)
+  {
+    return operator()(global_index);
+  }
+
+
+
+  template <typename Number>
+  template <typename OtherNumber>
+  inline
+  void ReadWriteVector<Number>::extract_subvector_to (const std::vector<size_type> &indices,
+                                                      std::vector<OtherNumber> &values) const
+  {
+    for (size_type i = 0; i < indices.size(); ++i)
+      values[i] = operator()(indices[i]);
+  }
+
+
+
+  template <typename Number>
+  template <typename ForwardIterator, typename OutputIterator>
+  inline
+  void ReadWriteVector<Number>::extract_subvector_to (ForwardIterator          indices_begin,
+                                                      const ForwardIterator    indices_end,
+                                                      OutputIterator           values_begin) const
+  {
+    while (indices_begin != indices_end)
+      {
+        *values_begin = operator()(*indices_begin);
+        indices_begin++;
+        values_begin++;
+      }
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number
+  ReadWriteVector<Number>::local_element (const size_type local_index) const
+  {
+    AssertIndexRange (local_index, this->size());
+
+    return val[local_index];
+  }
+
+
+
+  template <typename Number>
+  inline
+  Number &
+  ReadWriteVector<Number>::local_element (const size_type local_index)
+  {
+    AssertIndexRange (local_index, this->size());
+
+    return val[local_index];
+  }
+
+
+
+  template <typename Number>
+  inline
+  ReadWriteVector<Number> &
+  ReadWriteVector<Number>::operator = (const Number s)
+  {
+    std::fill(this->begin(), this->end(), s);
```

Comment:
Don't we allow only `s==0` in the current implementation? If so, can you copy the corresponding assertion?